### ISSUE TITLE
Fail to run in macOS 19

### ISSUE THREAD

**Issue by peekxc** [bloomberg/memray#614]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I would like to use `memray`, however it doesn't work *at all* on my system. 
```python 
# test_memray.py
import numpy as np 
x = np.zeros(100)
```
Running this file with the command `python -m memray run test_memray.py` yields an immediate seg. fault: 

```bash
Writing profile results into memray-test_memray.py.44886.bin
zsh: segmentation fault  python -m memray run test_memray.py
```
Though a file does get written, no recording is done; the result of `memray summary memray*.bin`  yields: 

```bash
⠋ Calculating high watermark... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   Memray ERROR: Failed to process memory record
⠋ Processing allocation records... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--Memray ERROR: Failed to process memory record
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃                                                           ┃     <Total ┃      Total ┃            ┃ Own Memory ┃ Allocati… ┃
┃ Location                                                  ┃    Memory> ┃   Memory % ┃ Own Memory ┃          % ┃     Count ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━┩
└───────────────────────────────────────────────────────────┴────────────┴────────────┴────────────┴────────────┴───────────┘
```

 I tried the previous versions of memray as well, at least back to 1.9.0, and none of the versions seem to work

### Expected Behavior

Any of the demonstrated behavior in the documentation

### Steps To Reproduce

1. Make any python file 
2. Run `memray run ...` on the python file
3. Observe the seg fault

### Memray Version

1.12.0

### Python Version

3.11

### Operating System

macOS

### Anything else?

======================================== System Information ========================================
System: Darwin
Node Name: MacBook-Pro-8.local
Release: 19.6.0
Version: Darwin Kernel Version 19.6.0: Tue Oct 12 18:34:05 PDT 2021; root:xnu-6153.141.43~1/RELEASE_X86_64
Machine: x86_64
Processor: i386

**Comment by pablogsal** [comment]
I think you have something weird in your system as we run the memray test suite on Intel Macs and Apple silicon macs without any problem.

Can you try to run memray under lldb and check where is segfaulting?

**Comment by pablogsal** [comment]
Also, can you tell us a bit more about your system?

What macOS version are you using? What hardware are you using?

**Comment by pablogsal** [comment]
> Processor: i386

What does this mean? Are you running a 32 bit macOS?

**Comment by peekxc** [comment]
> Also, can you tell us a bit more about your system?
> 
> What macOS version are you using? What hardware are you using?

10.15.7 (Catalina)

> What does this mean? Are you running a 32 bit macOS?

No sorry, idk why this was generated when I queried the processor. It's 64 bit macOS from 2019. 

> Can you try to run memray under lldb and check where is segfaulting?

This probably doesn't help, but I changed the source to: 

```python
import numpy as np 
import memray

with memray.Tracker("output_file.bin"):
  x = np.zeros(100)
```

Then executed it with the following outputs:
```{bash} 
(spri) mpiekenbrock{~/memray}% lldb python test_memray.py
(lldb) target create "python"
Current executable set to 'python' (x86_64).
(lldb) settings set -- target.run-args  "test_memray.py"
(lldb) run
Process 49064 launched: '/Users/mpiekenbrock/opt/miniconda3/envs/spri/bin/python' (x86_64)
Process 49064 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x0000000000000000
error: memory read failed for 0x0
Target 0: (python) stopped.
(lldb) quit
```

Again running the code normally just seg-faults: 

```bash
(spri) mpiekenbrock{~/laplacian_kernel/examples}% python test_memray.py
zsh: segmentation fault  python test_memray.py
```

I'm not entirely sure how to track down the segfault

**Comment by pablogsal** [comment]
When you see this in lldb:

error: memory read failed for 0x0

type “bt” and paste the output here

**Comment by peekxc** [comment]
Ahh it looks like I might need to update to OS X 10.14. 

I failed to compile memray from source using gcc-12 and clang-15. The relevant compiler error(s) in question: 

```{bash}
 In file included from src/memray/_memray.cpp:1361:
    src/memray/_memray/snapshot.h:110:59: error: 'value' is unavailable: introduced in macOS 10.14
                const auto& intersection = maybe_intersection.value();
                                                              ^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/optional:953:27: note: 'value' has been explicitly marked unavailable here
        constexpr value_type& value() &
                              ^
    In file included from src/memray/_memray.cpp:1377:
    src/memray/_memray/tracking_api.h:222:19: error: 'value' is unavailable: introduced in macOS 10.14
                trace.value().fill(1);
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/optional:953:27: note: 'value' has been explicitly marked unavailable here
        constexpr value_type& value() &
                              ^
...
```

For full reference, here's my compiler config: 

```{bash}
(venv) mpiekenbrock{~/memray}% g++ --version
g++-12 (Homebrew GCC 12.2.0) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

(venv) mpiekenbrock{~/memray}% clang++ --version
Homebrew clang version 15.0.7
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /usr/local/opt/llvm/bin
```

**Comment by pablogsal** [comment]
> I failed to compile memray from source using gcc-12 and clang-15. The relevant compiler error(s) in question:

Looks like your XCODE installation is broken somehow (which you can confirm by looking up those error messages)

**Comment by pablogsal** [comment]
Try setting `$ export MACOSX_DEPLOYMENT_TARGET=10.14`

**Comment by pablogsal** [comment]
You need to figure out how to fix that compilation problem because we cannot reproduce it on our side nor in our CI, so getting it to build on your system it's our only option at debugging this.

**Comment by pablogsal** [comment]
> Ahh it looks like I might need to update to OS X 10.14.

But you said you are using 10.15.7 (Catalina) no? I think you just need to export `MACOSX_DEPLOYMENT_TARGET` as I mentioned before

**Comment by peekxc** [comment]
> Try setting $ export MACOSX_DEPLOYMENT_TARGET=10.14

I did follow this originally, actually, though I thought I was meant to replace 10.14 with my OS's version; exporting thing does get past that error.

> You need to figure out how to fix that compilation problem because we cannot reproduce it on our side nor in our CI, so getting it to build on your system it's our only option at debugging this.

Yeah, that's fair. I might end up updating my OS to macOS 14 just to see if that fixes it. The current error 

```{bash}
    src/memray/_memray/macho_shenanigans.cpp:299:17: error: use of undeclared identifier '_dyld_shared_cache_contains_path'
                if (_dyld_shared_cache_contains_path(image_name) && strcmp(section->segname, SEG_TEXT) == 0)
                    ^
1 warning and 1 error generated.
```

Which I guess may be related to [this](https://stackoverflow.com/questions/68444867/checking-for-library-existence-on-macos-big-sur); will check back in once I try to fix the source.

**Comment by pablogsal** [comment]
Oh, that more or less confirms my guess: that function is not available at runtime in your system.

As a quick test can you substitute `_dyld_shared_cache_contains_path(image_name)` with `0`:


```
                if (0 && strcmp(section->segname, SEG_TEXT) == 0)

```

and check if it works?

**Comment by pablogsal** [comment]
Leave the issue open for now as we still need to fix it on our side

**Comment by pablogsal** [comment]
> It works :), everything seems to be fine now, have run a couple benchmarks to be sure.
> 
> Closing this issue. Thank you for your expediency

Awesome! If you want to use your version I suggest they you recompile memeray without the debug options so you don’t run a slow version

**Comment by pablogsal** [comment]
Fixed by #615
